### DISCUSSION TITLE
EPW modified source for the ~Quantum Espresso v7.0~ suite

### DISCUSSION THREAD

**Discussion by qiuzhanjun** [nakib/elphbolt#73] [General]
I am using the newest QE v7.0 and EPW v5.4. I have modified the EPW source code files according to the files of directory EPW/src/ but failed, due to the changes of ndegen_g. Is it possible to provide the modified EPW v5.4 code after the revision of this version?

**Comment by nakib** [comment]
Hi @qiuzhanjun. Thanks for using elphbolt and for your question! Would you be able to provide more information? For example, did you replace the whole file ephwann_shuffle.f90 of your version of EPW with the one provided with elphbolt? If yes, could you please try for this file adding just the lines 417-458 from elphbolt's version to your version of EPW right below line 389? Let me know if this solves your issue.

Best,
Nakib

**Comment by qiuzhanjun** [comment]
I appreciate your answer. According your advice, I recompiled the epw. Fortunately, it worked and didn't report an error. Next I'll test if it can print out the required files. I will give the feedback later.

**Comment by suan12** [comment]
Thanks for your reply@nakib. I know how to revise ephwann_shuffle.f90 according to the one provided with elphbolt.
However, I do not know how to modify rigid_epw.f90. THe source of rigid_epw.f90 in EPW v5.4 changes too much.
Is it possible to provide it for this version?
By the way, does the code take 4 phonons into considerations?

Best,
Pengfei Liu

**Reply by nakib** [reply]
Dear Pengfei Liu,

I took a quick look at the file `rigid_epw.f90` of the `EPW` module shipped with `QE` version 7.0 and 7.1. For these versions, I don't think this file needs to be replaced for generating results for `elphbolt`. That said, it is always a good idea to compare the results of the band structure, phonon dispersions, and matrix elements plotted along some path between `EPW` and `elphbolt`. I hope this helps!

Related to the above, the next major version of elphbolt will likely be able to work with a new version of `QE`.

Regarding 4-phonons, yes I plan to interface with an external 4-phonon calculator, possibly the `FourPhonon` code (https://github.com/FourPhonon/FourPhonon). The tentative idea is to generate 4-phonon scattering rates on a relatively coarse q-mesh using this external tool. Then I will interpolate these onto a fine q-mesh mesh inside `elphbolt`. The linear interpolator already exists in `elphbolt`. I don't have a timeline for when this might become available. I always appreciate any contribution from the community though. So if you have time and human-power to spend on this project to make it happen quicker, let me know! :)

Best regards,
Nakib

**Reply by suan12** [reply]
Thank you for your reply. However, I have the same problem above in ephwann_shuffle.f90. 
In QE v7.0, ndegen_g is defined as below:
INTEGER, ALLOCATABLE :: ndegen_g(:, :, :) 

In the one provided with elphbolt, it is shown as below:
INTEGER, ALLOCATABLE :: ndegen_g (:,:,:,:).

Thus, an error occurred during compilation:
"ephwann_shuffle.f90(427): error #6351: The number of subscripts is incorrect.   [NDEGEN_G]"

Best,
pfliu

**Reply by nakib** [reply]
I see. Thanks a lot for bringing this to my attention. Indeed the structure of `ndegen_g` has changed in `EPW` of `QEv7` onward. One can see this by  looking at this tensor in the subroutine `ephwannblochp` of `wan2bloch.f90`, for example.

I will need some time to figure out what is happening. In the mean time, you can continue using up to `QEv6.7MaX-Release`.

Best,
Nakib

**Reply by suan12** [reply]
Thanks. I am looking forward to your good news.
Best,
pfliu
